pkg/util/debouncer/debouncer.go

```diff
@@ -0,0 +1,279 @@
+//
+// A debouncer helps optimize expensive operations by:
+// 1. Grouping identical events that occur in rapid succession
+// 2. Processing each unique key only once after waiting periods expire
+//
+// Parameters:
+//   - BufferSize: Maximum number of pending events to buffer
+//   - KeyFunc: Function that extracts/defines the key from an input
+//   - MinWait: Cooldown period after receiving an event. If another event with the
+//     same key arrives during this period, the timer resets and we wait another MinWait duration.
+//   - MaxWait: Maximum time any event will wait before processing. Even if new events
+//     for the same key keep arriving, we guarantee processing after MaxWait from the first event.
+//
+// Example usage:
+//
+//	debouncer := NewDebouncer(DebouncerOpts[string]{
+//		BufferSize: 10,
+//		KeyFunc:    func(s string) string { return s },
+//		MinWait:    time.Second * 10,
+//		MaxWait:    time.Minute,
+//	})
+//
+//	debouncer.Start(ctx, func(key string) error {
+//		// This is where you perform the expensive operation
+//		return doSuperExpensiveCommand(key)
+//	})
+//
+//	// Queue events
+//	debouncer.Add("user-1")
+//	// Adding the same key resets MinWait but not MaxWait
+//	debouncer.Add("user-1")
+//
+// The event will be processed when either MinWait expires (after the most recent add)
+// or MaxWait expires (after the first add), whichever comes first.
+func NewDebouncer[T any](opts DebouncerOpts[T]) *Debouncer[T] {
+	if opts.BufferSize <= 0 {
+		opts.BufferSize = 100
+	}
+	if opts.MinWait <= 0 {
+		opts.MinWait = time.Minute
+	}
+	if opts.MaxWait <= 0 {
+		opts.MaxWait = 5 * time.Minute
+	}
+	if opts.KeyFunc == nil {
+		panic("KeyFunc is required")
+	}
+
+	if opts.Reg == nil {
+		opts.Reg = prometheus.NewRegistry()
+	}
+
+	return &Debouncer[T]{
+		opts:         opts,
+		buffer:       make(chan T, opts.BufferSize),
+		debouncers:   make(map[string]*keyDebouncer[T]),
+		metrics:      newMetrics(opts.Reg, opts.Name),
+		errorHandler: opts.ErrorHandler,
+	}
+}
+
+func (d *Debouncer[T]) Add(value T) error {
+	select {
+	case d.buffer <- value:
+		d.metrics.itemsAddedCounter.Inc()
+		return nil
+	default:
+		d.metrics.itemsDroppedCounter.Inc()
+		return ErrBufferFull
+	}
+}
+
+func (d *Debouncer[T]) Start(ctx context.Context, processFunc ProcessFunc[T]) {
+	d.ctx, d.cancel = context.WithCancel(ctx)
+	d.wg.Add(1)
+	go func() {
+		defer d.wg.Done()
+		for {
+			select {
+			case <-d.ctx.Done():
+				return
+			case value := <-d.buffer:
+				d.processValue(value, processFunc)
+			}
+		}
+	}()
+}
+
+func (d *Debouncer[T]) Stop() {
+	if d.cancel != nil {
+		d.cancel()
+		d.wg.Wait()
+	}
+}
+
+func (d *Debouncer[T]) SetMinWait(minWait time.Duration) {
+	d.opts.MinWait = minWait
+}
+
+func (d *Debouncer[T]) SetMaxWait(maxWait time.Duration) {
+	d.opts.MaxWait = maxWait
+}
+
+func (d *Debouncer[T]) processValue(value T, processFunc ProcessFunc[T]) {
```

Comment:
Great documentation overall. It would be helpful to have a bit more within this function.